torch/csrc/jit/script/compiler.cpp

```diff
@@ -836,22 +912,107 @@ struct to_ir {
     return emitIfExpr(expr.range(), cond_value, true_expr, false_expr);
   }
 
+  BoolInfo mergeAnd(BoolInfo a, BoolInfo b) {
+    // if the result of an AND is true, both a & b had to be true,
+    // so we take the union of a.true_refinements and b.true_refinements.
+    // if the result is false, either a or b could have been false,
+    // so we take their intersection.
+    auto true_refinements =
+        unionRefinements(a.true_refinements, b.true_refinements);
+    auto false_refinements =
+        intersectRefinements(a.false_refinements, b.false_refinements);
+    return BoolInfo(true_refinements, false_refinements);
+  }
+
+  BoolInfo mergeOr(BoolInfo a, BoolInfo b) {
+    // if the result of an OR is true, either a & b could have been true,
+    // so we take the intersection of a.true_refinements & b.true_refinements.
+    // if the result is false, both a and b had to be false,
+    // so we take their union.
+    auto true_refinements =
+        intersectRefinements(a.true_refinements, b.true_refinements);
+    auto false_refinements =
+        unionRefinements(a.false_refinements, b.false_refinements);
+    return BoolInfo(true_refinements, false_refinements);
+  }
+
+  // return the intersection of the values to type mappings in a and b whose
+  // types can be unified
+  Refinements intersectRefinements(Refinements a, Refinements b) {
+    Refinements ret;
+    for (auto& value_type : a) {
+      Value* v_1 = value_type.first;
+      TypePtr t_1 = value_type.second;
+      auto maybe_t_2 = b.find(v_1);
+      if (maybe_t_2 != b.end()) {
+        TypePtr t_2 = maybe_t_2->second;
+        auto maybe_unified_type = unifyTypes(t_1, t_2);
+        if (maybe_unified_type)
+          ret[v_1] = *maybe_unified_type;
+      }
+    }
+    return ret;
+  }
+
+  // return the union of the values to type mappings in a and b whose
+  // types can be unified
+  Refinements unionRefinements(Refinements a, Refinements b) {
+    Refinements ret;
+    for (auto& value_type : a) {
+      Value* v_1 = value_type.first;
+      TypePtr t_1 = value_type.second;
+      auto maybe_t_2 = b.find(v_1);
+      if (maybe_t_2 != b.end()) {
+        TypePtr t_2 = maybe_t_2->second;
+        auto maybe_unified_type = unifyTypes(t_1, t_2);
```

Comment:
This seems wrong. Lets say you have two refinements known to be true:

a -> A
a -> B
with A <: B

Then you known 'a -> A' but unify(A, B) == B. It is possible you cannot cause this to fail yet because we never introduce a refinement with this structure yet, but it shouldn't be left incorrect. I want to avoid adding a typeMeet operator because it is a burden to maintain, but you can choose t_1 if t_1 <: t_2, t_2 if t_2 <: t_1, and drop the refinement if their is true.